File: packages/vexnor/src/execution/audit-log-plugin.ts
import type { SqlQueryPipelinePlugin, SqlPipelineEndArgs } from "./sql-query-pipeline-plugin.js";

export type AuditLogArgs<TContext extends Record<string, unknown> = Record<string, unknown>> =
   SqlPipelineEndArgs<TContext>;

export type AuditLogPluginOptions<TContext extends Record<string, unknown> = Record<string, unknown>> = {
   /**
    * Name identifying this plugin instance.
    * Defaults to `"AuditLogPlugin"`.
    */
   name?: string;

   /**
    * Opt-in projection of context into the audit log.
    * When set, `AuditLogArgs.context` contains the resolver's return value.
    * When absent, `context` is `null` and raw context is never logged.
    */
   contextLogResolver?(context: TContext): Record<string, unknown>;

   /** Called after every query execution with the trimmed audit log args. */
   onLog: (args: AuditLogArgs<TContext>) => void;
};

/**
 * Built-in audit log plugin. Attach via `pipeline.use(new AuditLogPlugin({ onLog: ... }))`.
 *
 * Fires in the `end()` phase — after every pipeline execution (success, failure, or rejection).
 * Trims context via `contextLogResolver` before passing to `onLog`. Raw context is never forwarded.
 */
export class AuditLogPlugin<
   TContext extends Record<string, unknown> = Record<string, unknown>,
> implements SqlQueryPipelinePlugin<TContext> {
   readonly name: string;

   constructor(public readonly options: AuditLogPluginOptions<TContext>) {
      this.name = options.name ?? "AuditLogPlugin";
   }

   end(args: SqlPipelineEndArgs<TContext>): void {
      const context = this.options.contextLogResolver?.(args.context) ?? null;
      const log = { ...args, context: context as TContext };
      this.options.onLog(log);
   }
}

File: stacks/dotnet/src/Vexnor.Core/Execution/AuditLogPlugin.cs
namespace Vexnor.Core.Execution;

/// <summary>
/// Audit log entry emitted after every pipeline execution.
/// </summary>
public sealed class AuditLogEntry
{
    public required string Name { get; init; }
    public required string Hash { get; init; }
    public required string? Location { get; init; }
    public required long DurationMs { get; init; }
    public required string? Error { get; init; }
    public required Dictionary<string, object?>? Context { get; init; }
}

/// <summary>
/// Configuration for the audit log plugin.
/// </summary>
public sealed class AuditLogOptions
{
    /// <summary>Plugin name. Defaults to "AuditLogPlugin".</summary>
    public string Name { get; init; } = "AuditLogPlugin";

    /// <summary>
    /// Opt-in projection of context into the audit log.
    /// Raw context is never forwarded — only what this resolver returns.
    /// When null, Context in the log entry is null.
    /// </summary>
    public Func<Dictionary<string, object?>, Dictionary<string, object?>>? ContextLogResolver { get; init; }

    /// <summary>Called after every pipeline execution (success, failure, or rejection).</summary>
    public required Action<AuditLogEntry> OnLog { get; init; }
}

/// <summary>
/// Built-in audit log plugin. Fires in End() — after every execution including denials.
/// </summary>
public sealed class AuditLogPlugin : IQueryPipelinePlugin
{
    private readonly AuditLogOptions _options;

    public string Name => _options.Name;

    public AuditLogPlugin(AuditLogOptions options)
    {
        _options = options;
    }

    public void End(PipelineEndArgs args)
    {
        var context = _options.ContextLogResolver?.Invoke(args.Execution.Context);

        _options.OnLog(new AuditLogEntry
        {
            Name = args.Execution.Name,
            Hash = args.Execution.Hash,
            Location = args.Execution.Location,
            DurationMs = args.DurationMs,
            Error = args.Error?.Message,
            Context = context,
        });
    }
}
